### PR TITLE
HPCC-29192 Avoid path traversal when accessing dropzone and log file

### PR DESCRIPTION
**Pull request by wangkx**
1. The file name check is added When downloading a dropzone file.
2. The path check is added When listing log files or dropzone files.
3. The path/file checks are added when spraying/despraying dropzone file.
4. The path/file checks are added in the DropZoneFiles and DeleteDropZoneFiles.

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/services/ws_fs/ws_fsService.cpp [hunk 1]
```diff
@@ -3444,6 +3459,9 @@ bool CFileSprayEx::onDeleteDropZoneFiles(IEspContext &context, IEspDeleteDropZon
             if (!file || !*file)
                 continue;
 
+            if (findPathSepChar(file))
```
Comment (afishbeck) [inline esp/services/ws_fs/ws_fsService.cpp:3462]: Can we move this parameter validation to the top, just below 
```
 if (!files.ordinality())
            throw MakeStringException(ECLWATCH_INVALID_INPUT, "File not specified.");
```
In a separate small loop
```
ForEachItemIn(idx, files)
{
    if (findPathSepChar(file))
        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid file name %s", file);
}
```
That would make it so if any filenames were bad we wouldn't delete some but not all.

Comment (wangkx) [reply esp/services/ws_fs/ws_fsService.cpp]: I just changed. Please review.

File: esp/services/ws_fs/ws_fsService.cpp [hunk 2]
```diff
@@ -3407,10 +3419,19 @@ bool CFileSprayEx::onDeleteDropZoneFiles(IEspContext &context, IEspDeleteDropZon
...
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid path %s", directory);
+
         const char* osStr = req.getOS();
         StringArray & files = req.getNames();
         if (!files.ordinality())
             throw MakeStringException(ECLWATCH_INVALID_INPUT, "File not specified.");
+        ForEachItemIn(idx, files)
+        {
+            const char* file = files.item(idx);
+            if (findPathSepChar(file))
```
Comment (ghalliday) [inline esp/services/ws_fs/ws_fsService.cpp:3432]: Is this too strict?  Should it be containsRelPaths()?

File: esp/services/ws_fs/ws_fsBinding.cpp [hunk 3]
```diff
@@ -399,6 +399,8 @@ int CFileSpraySoapBindingEx::downloadFile(IEspContext &context, CHttpRequest* re
         request->getParameter("Path", pathStr);
         request->getParameter("Name", nameStr);
         request->getParameter("DropZoneName", dropZoneName);
+        if (findPathSepChar(nameStr))
```
Comment (ghalliday) [inline esp/services/ws_fs/ws_fsBinding.cpp:402]: Is this also too strict?

**Review by ghalliday — COMMENTED** [review]
@wangkx thanks for the quick response.  A couple of questions whether the changes are too strict.

**Comment by ghalliday** [comment]
Comments from Tony Kirk suggest that they do spray and despray from sub-directories on the landing zone, and there isn't any reason currently why that wouldn't work.  Safest to be less restrictive and only check for relative paths.  We can always tighten up further later.

**Review by afishbeck — APPROVED** [review]

**Review by ghalliday — APPROVED** [review]
I think we should consider also applying this to older builds, but I will merge to 9.0.x first.

**Pull request merged**